### PR TITLE
Isolated tests simplified

### PR DESCRIPTION
**Pull request by msgroi**

### CODE REVIEW

File: src/test/java/com/salesforce/dynamodbv2/PutTest.java [hunk 1]
```diff
@@ -74,12 +76,12 @@ void putOverwrite(TestArgument testArgument) {
...
         });
     }
 
-    @TestTemplate
-    @ExtendWith(TestTemplateWithDataSetup.class)
+    @ParameterizedTest(name = "{arguments}")
+    @ArgumentsSource(DefaultArgumentProvider.class)
     void putHkRkTable(TestArgument testArgument) {
         testArgument.getOrgs().forEach(org -> {
             MT_CONTEXT.setContext(org);
```
Comment (busjaeger) [inline src/test/java/com/salesforce/dynamodbv2/PutTest.java]: Maybe worth considering an `MT_CONTEXT.withContext(org, () -> ...)` that sets the context to the given org and upon completion sets the context back to the previous value (which is likely null in most cases). To collapse these two lines, there could also be a `testArgument.getOrgs().withEachContext(() -> ...)` that iterates and calls that method for each loop.

File: src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java [hunk 2]
```diff
@@ -111,18 +114,7 @@ public TestArgumentSupplier(AmazonDynamoDB rootAmazonDynamoDb) {
          * byAccount
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java]: re line 104 above ("return ImmutableList.of(S, N, B);"): consider "return Arrays.asList(ScalarAttributeType.values());"

File: src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java [hunk 3]
```diff
@@ -41,24 +45,23 @@
...
+ * Builds of a list of Argument objects, where each Argument is a 3 element Object[] array, where the first element
  * is the AmazonDynamoDB instance being tested and the second element is a list of orgs that have been designated
- * to be used when testing that instance.
+ * to be used when testing that instance, and the third is the attribute type of the hash key of the table to be
+ * tested.
  *
  * @author msgroi
  */
-class TestArgumentSupplier implements Supplier<List<Arguments>> {
+public class ArgumentBuilder implements Supplier<List<Arguments>> {
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java]: have get() return "List<TestArgument>" instead of "List<Arguments>"?

Comment (horwitzj) [reply src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java]: that should be "List&lt;TestArgument&rt;"

File: src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java [hunk 4]
```diff
@@ -111,18 +114,7 @@ public TestArgumentSupplier(AmazonDynamoDB rootAmazonDynamoDb) {
...
-
-                @Override
-                public void shutdown() {
-                    getServers().forEach(LocalDynamoDbServer::stop);
-                }
-            })
+            .withAccountMapper(new DynamicAccountMtMapper())
             .withContext(MT_CONTEXT).build();
 
         /*
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/ArgumentBuilder.java]: perhaps leave a quick comment that 'endsWith("3")' is an example of a zany, arbitrary rule

File: src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    public DefaultArgumentProvider(ArgumentBuilder argumentBuilder, TestSetup testSetup) {
+        this.argumentBuilder = argumentBuilder;
+        this.testSetup = testSetup;
+    }
+
+    @Override
+    public Stream<? extends Arguments> provideArguments(ExtensionContext extensionContext) {
+        List<Arguments> arguments = argumentBuilder.get();
+        arguments.forEach(arguments1 -> testSetup.setupTest((TestArgument) arguments1.get()[0]));
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java]: ... update when JUnit allows access to test parameters...

File: src/test/java/com/salesforce/dynamodbv2/testsupport/IsolatedArgumentProvider.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                amazonDynamoDb = AmazonDynamoDBClientBuilder.standard()
+                    .withRegion(ArgumentBuilder.REGION).build();
+                amazonDynamoDbStreams = AmazonDynamoDBStreamsClientBuilder.standard()
+                    .withRegion(ArgumentBuilder.REGION).build();
+            }
+            initialized = true;
+        }
+    }
+
+    public static AmazonDynamoDB getAmazonDynamoDb() {
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/IsolatedArgumentProvider.java]: maybe rename to (something like) "getAndInitializeDynamoDb"? (sim. w/getAmazonDynamoDbStreams)

File: src/test/java/com/salesforce/dynamodbv2/PutTest.java (hunk 1 above)
Comment (msgroi) [reply src/test/java/com/salesforce/dynamodbv2/PutTest.java]: Will do.

File: src/main/java/com/salesforce/dynamodbv2/mt/context/impl/MtAmazonDynamoDbContextProviderImpl.java [hunk 7]
```diff
@@ -37,6 +38,20 @@ public String getContext() {
         }
     }
 
+    @Override
+    public void withContext(String org, Runnable runnable) {
+        String origContext = getContext();
+        try {
+            setContext(org);
+            runnable.run();
+        } catch (Exception e) {
```
Comment (busjaeger) [inline src/main/java/com/salesforce/dynamodbv2/mt/context/impl/MtAmazonDynamoDbContextProviderImpl.java]: Why catch Exception here?

File: src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java [hunk 8]
```diff
@@ -42,9 +80,19 @@ public DefaultArgumentProvider(ArgumentBuilder argumentBuilder, TestSetup testSe
...
     @Override
     public Stream<? extends Arguments> provideArguments(ExtensionContext extensionContext) {
-        List<Arguments> arguments = argumentBuilder.get();
-        arguments.forEach(arguments1 -> testSetup.setupTest((TestArgument) arguments1.get()[0]));
-        return arguments.stream();
+        List<TestArgument> arguments = argumentBuilder.get();
+        /*
+         * Before returning the list of Arguments back to JUnit so it can pass each in as a test invocation, we
+         * are calling TestSetup.setupTest().  This provides a hook for doing additional data setup for each
+         * parameterized test argument.  The default TestSetup implementation, DefaulTestSetup, creates a default
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java]: DefaulTestSetup -> DefaultTestSetup (i.e., add a "t")

File: src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java [hunk 9]
```diff
@@ -42,9 +80,19 @@ public DefaultArgumentProvider(ArgumentBuilder argumentBuilder, TestSetup testSe
...
+         * set of tables and populates them with a default set of data.  This begs the question, why not declare a
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java]: "data.  This begs the question," -> "data, which leads to the question,"

File: src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java [hunk 10]
```diff
@@ -42,9 +80,19 @@ public DefaultArgumentProvider(ArgumentBuilder argumentBuilder, TestSetup testSe
...
+         * standard JUnit BeforeEachCallback handler and perform the data set up there.  The answer is that the
+         * ExtensionContext provided in JUnit callback methods do not have access to @ParameterizedTest arguments.
+         * Note that this is a known deficiency that's currently slated to be addressed in an upcoming version of
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java]: "that's" -> "which is"

File: src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java [hunk 11]
```diff
@@ -8,14 +8,52 @@
...
- * a @ParameterizedTest's @ArgumentProvider annotation.
+ * Default implementation of the ArgumentProvider interface.  It is meant to be referenced in a @ParameterizedTest's
+ * &#064;ArgumentProvider annotation.
  *
  * <p>It delegates to the ArgumentBuilder to get a list of TestArguments.  Each TestArgument will be used as an input
  * to a test invocation.  Before returning the list of TestArgument's, it calls setTestSetup on its DefaultTestSetup.
  * The DefaultTestSetup implementation creates tables for each TestArgument's org/AmazonDynamoDB/hashKeyAttrType
  * combination.
  *
+ * <p>To use with the DefaultArgumentBuilder and DefaultTestSetup add the following annotation on your test ...
```
Comment (horwitzj) [inline src/test/java/com/salesforce/dynamodbv2/testsupport/DefaultArgumentProvider.java]: "test ..." -> "test..." (and other places)

File: src/main/java/com/salesforce/dynamodbv2/mt/context/impl/MtAmazonDynamoDbContextProviderImpl.java (hunk 7 above)
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/mt/context/impl/MtAmazonDynamoDbContextProviderImpl.java]: Fixed 0fed4666de842c083e0e06e45c9d5c2e635bd375

**Pull request merged**